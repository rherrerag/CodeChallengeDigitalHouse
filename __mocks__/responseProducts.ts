const responseOK = {
  data: [
    {
      createdAt: '2023-03-11T04:50:50.205Z',
      product: 'Recycled Plastic Tuna',
      points: 92984,
      image: 'https://loremflickr.com/640/480/technics',
      is_redemption: false,
      id: '2',
    },
    {
      createdAt: '2023-03-11T04:35:35.259Z',
      product: 'Fantastic Granite Bacon',
      points: 42416,
      image: 'https://loremflickr.com/640/480/technics',
      is_redemption: false,
      id: '3',
    },
    {
      createdAt: '2023-03-11T13:57:33.428Z',
      product: 'Fantastic Fresh Gloves',
      points: 23913,
      image: 'https://loremflickr.com/640/480/city',
      is_redemption: true,
      id: '4',
    },
    {
      createdAt: '2023-03-11T12:17:47.118Z',
      product: 'Rustic Rubber Bacon',
      points: 69814,
      image: 'https://loremflickr.com/640/480/people',
      is_redemption: true,
      id: '5',
    },
    {
      createdAt: '2023-02-04T07:48:16.249Z',
      product: 'Licensed Metal Salad',
      points: 93367,
      image: 'https://loremflickr.com/640/480/food',
      is_redemption: false,
      id: '12',
    },
  ],
  error: false,
  isCancel: false,
};

const responseError = {
  data: {message: 'Error', code: 'ERR'},
  error: true,
};

export {responseOK, responseError};
