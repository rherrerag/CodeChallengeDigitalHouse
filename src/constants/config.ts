const config = {
  FILTER: {
    MONTH: '3',
    YEAR: '2023',
  },
};

export default config;
