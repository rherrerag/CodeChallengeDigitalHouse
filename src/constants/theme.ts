import {extendTheme} from 'native-base';

const customTheme = extendTheme({
  colors: {
    primary: {
      100: '#334FFA',
      200: '#CFD6FF',
    },
    text: {
      title: '#020202',
      subtitle: '#9B9898',
      normal: '#000000',
    },
  },
});

export {customTheme};
