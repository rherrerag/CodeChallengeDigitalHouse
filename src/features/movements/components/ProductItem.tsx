import React, {memo} from 'react';
import {connect} from 'react-redux';
import {VStack} from 'native-base';
import {Movement} from '../interfaces/movement.interface';
import {TouchableOpacity} from 'react-native-gesture-handler';
import {useNavigation} from '@react-navigation/native';
import {StackNavigationProp} from '@react-navigation/stack';
import {monthNames} from '../../../common/utils/dateUtils';
import {
  MinusText,
  PlusText,
  ThumbnailImage,
  ItemContainer,
  PointValue,
  ProductTitle,
  DateSubtitle,
  ImageContainer,
  LeftContainer,
  RightContainer,
  ItemChevron,
} from './ProductItem.styled';

type ProductItemProps = {
  movement?: Movement | null;
};

export type RootStackParamList = {
  ProductDetail: {item: Movement | undefined | null};
};

const ProductItem: React.FC<ProductItemProps> = ({movement}) => {
  const navigation = useNavigation<StackNavigationProp<RootStackParamList>>();

  const goToProductDetail = (item: Movement | undefined | null) =>
    navigation.navigate('ProductDetail', {item: item});

  let fecha = new Date(movement?.createdAt);
  let fechaFormat: string = `${fecha.getDay()} de ${
    monthNames[fecha.getMonth()]
  }, ${fecha.getFullYear()}`;

  const renderMinus = () => <MinusText>-</MinusText>;
  const renderPlus = () => <PlusText>+</PlusText>;
  const sign = movement?.is_redemption ? renderMinus() : renderPlus();

  return (
    <TouchableOpacity onPress={() => goToProductDetail(movement)}>
      <ItemContainer space={2}>
        <ImageContainer>
          <ThumbnailImage source={{uri: movement?.image}} resizeMode="cover" />
        </ImageContainer>
        <LeftContainer>
          <VStack>
            <ProductTitle testID={movement?.product}>
              {movement?.product}
            </ProductTitle>
            <DateSubtitle>{fechaFormat}</DateSubtitle>
          </VStack>
        </LeftContainer>
        <RightContainer>
          <PointValue>
            {sign}
            {movement?.points.toLocaleString('es-MX')}
          </PointValue>
          <ItemChevron />
        </RightContainer>
      </ItemContainer>
    </TouchableOpacity>
  );
};

export default connect()(memo(ProductItem));
