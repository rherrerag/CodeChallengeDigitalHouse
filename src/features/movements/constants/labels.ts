const labels = {
  WELCOME: {
    MESSAGE: 'Bienvenido de Vuelta!',
    NAME: 'Ruben Rodriguez',
  },
  SUBTITLES: {
    POINTS: 'TUS PUNTOS',
    MOVEMENTS: 'TUS MOVIMIENTOS',
  },
  CARD: {
    POINTS: 'pts',
  },
  BUTTONS: {
    EARNED: 'Ganados',
    REDEEMED: 'Canjeados',
    ALL: 'Todos',
  },
  LOADING: {
    TITLE: 'Cargando',
  },
};

export default labels;
