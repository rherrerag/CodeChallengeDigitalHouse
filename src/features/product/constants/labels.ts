const labels = {
  PRODUCT_DETAIL: {
    SUBTITLE_DETAIL: 'Detalles del producto:',
    SUBTITLE_SUMMARY: 'Con esta compra acumulaste:',
    POINTS_TEXT: 'puntos',
    DATE_SUBSTRING: 'Comprado el',
  },
  ACCEPT_BUTTON: {
    TITLE: 'Aceptar',
  },
};

export default labels;
