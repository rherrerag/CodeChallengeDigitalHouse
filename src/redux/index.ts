import {configureStore, ThunkAction, Action} from '@reduxjs/toolkit';
import {movementsReducer} from '../features/movements/reducers/movements.reducer';

export const store = configureStore({
  reducer: {
    movements: movementsReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
