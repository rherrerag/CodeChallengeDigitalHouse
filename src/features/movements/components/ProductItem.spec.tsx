import React from 'react';
import {render, screen, waitFor} from '@testing-library/react-native';
import {responseOK} from '../../../../__mocks__/responseProducts';
import {Provider} from 'react-redux';
import ProductItem from './ProductItem';
import {store} from '../../../redux';
import mockFetch from 'jest-fetch-mock';
import {NativeBaseProvider} from 'native-base';
import {NavigationContainer} from '@react-navigation/native';
import {customTheme} from '../../../constants/theme';
import {Movement} from '../../movements/interfaces/movement.interface';

const movementMock: Movement = {
  createdAt: '2023-03-11T04:35:35.259Z',
  product: 'Fantastic Granite Bacon',
  points: 42416,
  image: 'https://loremflickr.com/640/480/technics',
  is_redemption: false,
  id: 3,
};

const inset = {
  frame: {x: 0, y: 0, width: 0, height: 0},
  insets: {top: 0, left: 0, right: 0, bottom: 0},
};

const ProductItemFactory = async () => {
  mockFetch.mockResponse(req => {
    if (req.url.includes('products')) {
      return Promise.resolve(JSON.stringify(responseOK.data));
    }

    return Promise.reject(new Error('not mapped url'));
  });
  render(
    <Provider store={store}>
      <NativeBaseProvider theme={customTheme} initialWindowMetrics={inset}>
        <NavigationContainer>
          <ProductItem movement={movementMock} />
        </NavigationContainer>
      </NativeBaseProvider>
    </Provider>,
  );
  await waitFor(() => expect(screen.getAllByTestId('movement-3')).toBeTruthy());
};

const mockedDispatch = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: jest.fn(),
      dispatch: mockedDispatch,
    }),
    useRoute: () => ({
      params: {
        item: movementMock,
      },
    }),
  };
});

describe('Product Detail Container', () => {
  beforeEach(() => {
    mockFetch.enableMocks();
    mockFetch.resetMocks();
    mockedDispatch.mockClear();
  });

  it('deberÃ­a generar snapshot de componente correctamente', async () => {
    await ProductItemFactory();
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
