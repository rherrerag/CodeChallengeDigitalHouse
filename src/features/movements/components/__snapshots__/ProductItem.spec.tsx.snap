// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Detail Container deber√≠a generar snapshot de componente correctamente 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNGestureHandlerButton
    collapsable={false}
    delayLongPress={600}
    enabled={true}
    exclusive={true}
    handlerTag={1}
    handlerType="NativeViewGestureHandler"
    onGestureEvent={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onHandlerStateChange={[Function]}
    rippleColor={0}
    testID="movement-3"
    touchSoundDisabled={false}
  >
    <View
      accessible={true}
      collapsable={false}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        dataSet={{}}
        style={
          [
            {
              "flexDirection": "row",
            },
            [
              {
                "flexDirection": "row",
                "paddingBottom": 4,
                "paddingTop": 4,
              },
              undefined,
            ],
          ]
        }
      >
        <View
          dataSet={{}}
          style={
            [
              {},
              [
                {
                  "alignSelf": "center",
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                },
                undefined,
              ],
            ]
          }
        >
          <Image
            resizeMode="cover"
            source={
              {
                "uri": "https://loremflickr.com/640/480/technics",
              }
            }
            style={
              [
                {
                  "borderBottomLeftRadius": 10,
                  "borderBottomRightRadius": 10,
                  "borderTopLeftRadius": 10,
                  "borderTopRightRadius": 10,
                  "height": 55,
                  "width": 55,
                },
                undefined,
              ]
            }
          />
        </View>
        <View
          dataSet={{}}
          style={
            {
              "width": 8,
            }
          }
        />
        <View
          dataSet={{}}
          style={
            [
              {},
              [
                {
                  "alignSelf": "center",
                  "flexBasis": 0,
                  "flexDirection": "row",
                  "flexGrow": 2,
                  "flexShrink": 1,
                  "justifyContent": "flex-start",
                },
                undefined,
              ],
            ]
          }
        >
          <View
            dataSet={{}}
            style={{}}
          >
            <Text
              dataSet={{}}
              style={
                [
                  {
                    "backgroundColor": undefined,
                    "color": "#171717",
                    "fontFamily": undefined,
                    "fontSize": 14,
                    "fontStyle": "normal",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                    "lineHeight": 21,
                    "textDecorationLine": undefined,
                  },
                  [
                    {
                      "fontSize": 14,
                      "fontWeight": "800",
                      "lineHeight": 19,
                      "paddingBottom": 7,
                    },
                    undefined,
                  ],
                ]
              }
            >
              Fantastic Granite Bacon
            </Text>
            <View
              dataSet={{}}
              style={{}}
            />
            <Text
              dataSet={{}}
              style={
                [
                  {
                    "backgroundColor": undefined,
                    "color": "#171717",
                    "fontFamily": undefined,
                    "fontSize": 14,
                    "fontStyle": "normal",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                    "lineHeight": 21,
                    "textDecorationLine": undefined,
                  },
                  [
                    {
                      "fontSize": 12,
                      "lineHeight": 16,
                    },
                    undefined,
                  ],
                ]
              }
            >
              6 de Marzo, 2023
            </Text>
          </View>
        </View>
        <View
          dataSet={{}}
          style={
            {
              "width": 8,
            }
          }
        />
        <View
          dataSet={{}}
          style={
            [
              {},
              [
                {
                  "alignSelf": "center",
                  "flexBasis": 0,
                  "flexDirection": "row",
                  "flexGrow": 1,
                  "flexShrink": 1,
                  "justifyContent": "flex-end",
                },
                undefined,
              ],
            ]
          }
        >
          <Text
            dataSet={{}}
            style={
              [
                {
                  "backgroundColor": undefined,
                  "color": "#171717",
                  "fontFamily": undefined,
                  "fontSize": 14,
                  "fontStyle": "normal",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                  "lineHeight": 21,
                  "textDecorationLine": undefined,
                },
                [
                  {
                    "alignSelf": "center",
                    "fontSize": 16,
                    "fontWeight": "800",
                    "justifyContent": "center",
                    "lineHeight": 22,
                  },
                  undefined,
                ],
              ]
            }
          >
            <Text
              dataSet={{}}
              style={
                [
                  {
                    "backgroundColor": undefined,
                    "fontFamily": undefined,
                    "fontSize": undefined,
                    "fontStyle": undefined,
                    "fontWeight": undefined,
                    "textDecorationLine": undefined,
                  },
                  [
                    {
                      "color": "green",
                      "fontSize": 16,
                      "fontWeight": "800",
                      "lineHeight": 22,
                    },
                    undefined,
                  ],
                ]
              }
            >
              +
            </Text>
            42,416
          </Text>
          <RNSVGSvgView
            accessibilityRole="image"
            align="xMidYMid"
            bbHeight={16}
            bbWidth={16}
            color={4285756275}
            dataSet={{}}
            focusable={false}
            meetOrSlice={0}
            minX={0}
            minY={0}
            stroke=""
            style={
              [
                {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                [
                  {
                    "color": "#737373",
                    "height": 16,
                    "width": 16,
                  },
                  [
                    {
                      "alignSelf": "center",
                      "fontSize": 16,
                      "lineHeight": 22,
                      "marginLeft": 4,
                      "marginTop": 2,
                    },
                    undefined,
                  ],
                ],
                {
                  "flex": 0,
                  "height": 16,
                  "width": 16,
                },
              ]
            }
            tintColor={4285756275}
            vbHeight={24}
            vbWidth={24}
          >
            <RNSVGGroup
              propList={
                [
                  "stroke",
                ]
              }
              stroke={null}
            >
              <RNSVGGroup>
                <RNSVGPath
                  d="M6 3.77141L14.3785 12.0027L6.00268 20.2314L7.80293 22L18 12L7.80293 2L6 3.77141Z"
                  fill={
                    [
                      2,
                    ]
                  }
                  propList={
                    [
                      "fill",
                      "stroke",
                    ]
                  }
                  stroke={null}
                />
              </RNSVGGroup>
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
      </View>
    </View>
  </RNGestureHandlerButton>
</RNCSafeAreaProvider>
`;
